cmd/xdpcap/filter.go

```diff
@@ -85,13 +90,30 @@ func newFilterWithMap(hookMap *ebpf.Map, opts filterOpts) (*filter, error) {
 		actions:  opts.actions,
 	}
 
-	for _, action := range opts.actions {
-		program, err := newProgram(opts.filter, action, perfMap)
+	xdpFragsMode := false
+	for i, action := range opts.actions {
+		program, err := newProgram(opts.filter, action, perfMap, xdpFragsMode)
 		if err != nil {
 			return nil, errors.Wrapf(err, "loading filter program for %v", action)
 		}
 
 		err = attachProg(hookMap, program.program.FD(), action)
+		if errors.Is(err, unix.EINVAL) && i == 0 {
+			// attempt to load first action in XDP frags mode and retry attaching
+			// if this doesn't work then there is some other error and the XDP frags
+			// mode is not the culprit
+			xdpFragsMode = true
+
+			program, err = newProgram(opts.filter, action, perfMap, xdpFragsMode)
+			if err != nil {
+				return nil, stdErr.Join(errors.Wrapf(err, "loading filter program in XDP frags mode for %v", action), err)
+			}
+
+			if retryErr := attachProg(hookMap, program.program.FD(), action); retryErr == nil {
```

Comment:
If there's a `retryErr`, I don't think we'll end up returning the original `attachProg()` error - the second call to `newProgram()` will overwrite `err`.

Reply:
gooc catch. I simplified a little bit the code over here and fixed it so that the original error doesn't get lost. I also reworded the commit message to use `XDP frags` everywhere and reflect the current state of the codebase